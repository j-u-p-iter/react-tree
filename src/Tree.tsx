import * as React from "react";

interface IChildrenProps {
  api: {
    toggleNode: Tree["toggleNode"];
    getTree: () => ITreeState["tree"];
  };
}

export type IConfigItem<T = {}> = T & {
  id: string;
  title: string;
  children?: IConfigItem[];
};

type TreeType = INode[];

interface INode extends IConfigItem {
  active: boolean;
  children: TreeType;
}

interface ITreeProps {
  config: IConfigItem[];
  activeId?: string;
  children: (childrenProps: IChildrenProps) => React.ReactNode;
}

interface ITreeState {
  tree: TreeType;
}

interface ITreeMap {
  [key: string]: any;
}

export class Tree extends React.Component<ITreeProps, ITreeState> {
  public state: ITreeState = {
    tree: []
  };
  private treeMap: ITreeMap = {};

  public toggleNode = (event: React.SyntheticEvent<HTMLElement>): void => {
    event.stopPropagation();

    this.extendOnToggle(event.currentTarget.id);
  };

  public componentDidMount() {
    this.initializeTree(this.props.activeId);
  }

  public render() {
    return this.props.children({
      api: {
        toggleNode: this.toggleNode,
        getTree: () => this.state.tree
      }
    });
  }

  private createTreeExtender = (
    customExtenderFn: (node: any) => Partial<INode>
  ) => {
    return function extender(data: IConfigItem[]): TreeType {
      return data.reduce((result: any, node: any) => {
        return [
          ...result,
          {
            ...node,
            ...customExtenderFn(node),
            children: node.children ? extender(node.children) : null
          }
        ];
      }, []);
    };
  };

  private setTreeState = (treeState: TreeType) => {
    this.setState({
      tree: treeState
    });
  };

  private initializeTree = (activeId?: INode["id"]): void => {
    this.treeMap = this.createTreeMap(this.props.config);

    const initialTreeState = this.createTreeExtender((node: IConfigItem) => ({
      active:
        activeId && this.treeMap[activeId].includes(node.id) ? true : false
    }))(this.props.config);

    this.setTreeState(initialTreeState);
  };

  private extendOnToggle = (nodeId: string): void => {
    const toggledTreeState = this.createTreeExtender((node: INode) => ({
      active: nodeId === node.id ? !node.active : node.active
    }))(this.state.tree);

    this.setTreeState(toggledTreeState);
  };

  private createTreeMap = (
    data: IConfigItem[],
    parentNodePath?: string
  ): ITreeMap => {
    return data.reduce((resultTreeMap: ITreeMap, node: IConfigItem) => {
      const nodePath = parentNodePath
        ? `${parentNodePath}:${node.id}`
        : node.id;

      resultTreeMap[node.id] = nodePath;

      if (node.children) {
        resultTreeMap = {
          ...resultTreeMap,
          ...this.createTreeMap(node.children, nodePath)
        };
      }

      return resultTreeMap;
    }, {});
  };
}
