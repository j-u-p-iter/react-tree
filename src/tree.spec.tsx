import * as React from 'react';

import { Tree } from './Tree'

import { render, fireEvent } from 'react-testing-library';


describe('Tree', () => {
  it('renders properly', () => {
    const treeEl = ( 
      <Tree activeId='/some-id-2' config={
        [
          { 
            id: '/some-id',
            title: 'some title',
            children: [
              { 
                id: '/some-new',
                title: 'cool title',
                children: [
                  { 
                    id: '/some-id-1',
                    title: 'coolio title',
                    children: [
                      { 
                        id: '/some-id-2',
                        title: 'some title',
                        children: [
                          { 
                            id: '/some-id-10',
                            title: 'some title',
                          },
                          { 
                            id: '/some-id-20',
                            title: 'some title',
                          }
                        ]
                      }
                    ]
                  }
                ]
              }
            ], 
          }, 
          {
            id: '/one-more-id',
            title: 'some title',
            children: [
              { 
                id: '/and-more-some-id',
                title: 'some title',
                children: [
                  { 
                    id: '/one-more-more-id',
                    title: 'some title',
                  }
                ]
              }
            ], 
          }
        ]
      }>
        {
          ({ api: { toggleState, getTree } }) => {
            const renderTree = (tree: any) => tree.map((node: any) => {
              return (
                <li data-path={node.path} data-active={node.active} key={node.id} id={node.id} onClick={toggleState}>
                  {node.title}
                  {node.children ? <ul>{renderTree(node.children)}</ul> : null}
                </li>
              );
            })

            return <ul>{renderTree(getTree())}</ul>;
          } 
        }
      </Tree>
    );

    const { container, getByText } = render(treeEl);

    expect(container.firstChild).toMatchSnapshot();

    fireEvent.click(getByText('coolio title'))

    expect(container.firstChild).toMatchSnapshot();
  })
});
