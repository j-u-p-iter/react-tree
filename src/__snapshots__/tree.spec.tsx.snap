// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree renders properly 1`] = `
<ul>
  <li
    data-active="true"
    id="/some-id"
  >
    some title
    <ul>
      <li
        data-active="true"
        id="/some-new"
      >
        cool title
        <ul>
          <li
            data-active="true"
            id="/some-id-1"
          >
            coolio title
            <ul>
              <li
                data-active="true"
                id="/some-id-2"
              >
                some title
                <ul>
                  <li
                    data-active="false"
                    id="/some-id-10"
                  >
                    some title
                  </li>
                  <li
                    data-active="false"
                    id="/some-id-20"
                  >
                    some title
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </li>
  <li
    data-active="false"
    id="/one-more-id"
  >
    some title
    <ul>
      <li
        data-active="false"
        id="/and-more-some-id"
      >
        some title
        <ul>
          <li
            data-active="false"
            id="/one-more-more-id"
          >
            some title
          </li>
        </ul>
      </li>
    </ul>
  </li>
</ul>
`;

exports[`Tree renders properly 2`] = `
<ul>
  <li
    data-active="true"
    id="/some-id"
  >
    some title
    <ul>
      <li
        data-active="true"
        id="/some-new"
      >
        cool title
        <ul>
          <li
            data-active="false"
            id="/some-id-1"
          >
            coolio title
            <ul>
              <li
                data-active="true"
                id="/some-id-2"
              >
                some title
                <ul>
                  <li
                    data-active="false"
                    id="/some-id-10"
                  >
                    some title
                  </li>
                  <li
                    data-active="false"
                    id="/some-id-20"
                  >
                    some title
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </li>
  <li
    data-active="false"
    id="/one-more-id"
  >
    some title
    <ul>
      <li
        data-active="false"
        id="/and-more-some-id"
      >
        some title
        <ul>
          <li
            data-active="false"
            id="/one-more-more-id"
          >
            some title
          </li>
        </ul>
      </li>
    </ul>
  </li>
</ul>
`;
